### PR TITLE
Trigger reparsing of the machete file on git repository changes

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

**Comment by PawelLipski** [comment]
Okay.... I already see that's (still) a bad idea, updating Git Machete status takes **a g e s**:

```
"AWT-EventQueue-0" #37 prio=6 os_prio=31 cpu=1419312.12ms elapsed=18315.26s tid=0x00007fc08703e200 nid=0x15703 runnable  [0x0000700009dfc000]
   java.lang.Thread.State: RUNNABLE
        at com.intellij.util.xml.impl.DomManagerImpl$2.shouldFireDomEvents(DomManagerImpl.java:135)
        at com.intellij.util.xml.impl.DomManagerImpl$2.prepareChange(DomManagerImpl.java:118)
        at com.intellij.openapi.vfs.newvfs.AsyncEventSupport.lambda$runAsyncListeners$1(AsyncEventSupport.java:90)
        at com.intellij.openapi.vfs.newvfs.AsyncEventSupport$$Lambda$3545/0x0000000802022248.run(Unknown Source)
        at com.intellij.openapi.application.ReadAction.lambda$run$1(ReadAction.java:59)
        at com.intellij.openapi.application.ReadAction$$Lambda$2366/0x00000008017ae670.compute(Unknown Source)
        at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:941)
        at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:68)
        at com.intellij.openapi.application.ReadAction.run(ReadAction.java:58)
        at com.intellij.openapi.vfs.newvfs.AsyncEventSupport.runAsyncListeners(AsyncEventSupport.java:90)
        at com.intellij.openapi.vfs.newvfs.AsyncEventSupport$1.before(AsyncEventSupport.java:54)
        at java.lang.invoke.LambdaForm$DMH/0x00000008003b9000.invokeInterface(java.base@17.0.5/LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/0x0000000800540000.invoke(java.base@17.0.5/LambdaForm$MH)
        at java.lang.invoke.LambdaForm$MH/0x0000000800540400.invokeExact_MT(java.base@17.0.5/LambdaForm$MH)
        at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:649)
        at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:625)
        at com.intellij.util.messages.impl.MessageBusImplKt.executeOrAddToQueue(MessageBusImpl.kt:470)
        at com.intellij.util.messages.impl.ToDirectChildrenMessagePublisher.publish$intellij_platform_core(CompositeMessageBus.kt:259)
        at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:425)
        at jdk.proxy2.$Proxy167.before(jdk.proxy2/Unknown Source)
        at com.intellij.util.FileContentUtilCore.lambda$reparseFiles$0(FileContentUtilCore.java:48)
        at com.intellij.util.FileContentUtilCore$$Lambda$3725/0x00000008020d54c0.run(Unknown Source)
        at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1023)
        at com.intellij.util.FileContentUtilCore.reparseFiles(FileContentUtilCore.java:37)
        at com.intellij.util.FileContentUtilCore.reparseFiles(FileContentUtilCore.java:28)
        at com.virtuslab.gitmachete.frontend.file.MacheteFileViewProviderFactory$MacheteFileViewProvider.lambda$subscribeToGitRepositoryChanges$0(MacheteFileViewProviderFactory.java:53)
        at com.virtuslab.gitmachete.frontend.file.MacheteFileViewProviderFactory$MacheteFileViewProvider$$Lambda$2471/0x00000008014eaaa8.run(Unknown Source)
        at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
        at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
        at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
        at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
        at com.intellij.openapi.application.impl.ApplicationImpl$3.run(ApplicationImpl.java:513)
        at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:75)
        at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:118)
        at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:42)
        at com.intellij.openapi.application.impl.FlushQueue$$Lambda$193/0x0000000800335c30.run(Unknown Source)
        at java.awt.event.InvocationEvent.dispatch(java.desktop/InvocationEvent.java:318)
        at java.awt.EventQueue.dispatchEventImpl(java.desktop/EventQueue.java:779)
        at java.awt.EventQueue$4.run(java.desktop/EventQueue.java:730)
        at java.awt.EventQueue$4.run(java.desktop/EventQueue.java:724)
        at java.security.AccessController.executePrivileged(java.base@17.0.5/AccessController.java:776)
        at java.security.AccessController.doPrivileged(java.base@17.0.5/AccessController.java:399)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(java.base@17.0.5/ProtectionDomain.java:86)
        at java.awt.EventQueue.dispatchEvent(java.desktop/EventQueue.java:749)
       ......
```

**Comment by PawelLipski** [comment]
Still, let's not give up on this one... it might be that a different IntelliJ API should be used in `subscribeToGitRepositoryChanges`

**Comment by PawelLipski** [comment]
Whoops the problem seems to be **not** in `FileContentUtilCore.reparseFiles` parsing more than just .git/machete... but instead `MacheteFileViewProvider` registering the listener multiple times, and then each of these listeners firing at the same time 😯

File: frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileUtils.java [hunk 1]
```diff
@@ -67,14 +66,17 @@ public static void saveDocument(PsiFile file) {
...
 
-  public static boolean macheteFileIsOpenedAndFocused(Project project, Path macheteFilePath) {
+  public static @Nullable VirtualFile getMacheteVirtualFileIfSelected(Project project) {
     val fileEditorManager = FileEditorManager.getInstance(project);
-    val macheteVirtualFile = List.of(fileEditorManager.getSelectedFiles())
-        .find(virtualFile -> virtualFile.getPath().equals(macheteFilePath.toString()));
-    if (macheteVirtualFile.isEmpty()) {
-      return false;
-    } else {
-      return fileEditorManager.getAllEditors(macheteVirtualFile.get()).length > 0;
```
Comment (PawelLipski) [inline frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileUtils.java:77]: I've checked experimentally... it looks that a selected file will always have at least editor present (that would intuitively make sense, at least)

File: frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java [hunk 2]
```diff
@@ -1,66 +1,23 @@
...
-
-  private static class MacheteFileViewProvider extends SingleRootFileViewProvider implements Disposable {
-
-    MacheteFileViewProvider(PsiManager manager,
-        VirtualFile virtualFile,
-        boolean eventSystemEnabled,
-        Language language) {
-      super(manager, virtualFile, eventSystemEnabled, language);
-      // TODO (#679): restore the subscription to git repo changes once we clarify why reparsing the files sometimes takes so long
-      // subscribeToGitRepositoryChanges(manager.getProject(), language)
```
Comment (PawelLipski) [inline frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java:39]: Declared in plugin.xml instead

Comment (mkondratek) [reply frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java:39]: 🚀

File: frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java [hunk 3]
```diff
@@ -1,66 +1,23 @@
...
   }
-
-  private static class MacheteFileViewProvider extends SingleRootFileViewProvider implements Disposable {
-
-    MacheteFileViewProvider(PsiManager manager,
-        VirtualFile virtualFile,
-        boolean eventSystemEnabled,
-        Language language) {
-      super(manager, virtualFile, eventSystemEnabled, language);
-      // TODO (#679): restore the subscription to git repo changes once we clarify why reparsing the files sometimes takes so long
```
Comment (mkondratek) [inline frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java:38]: https://github.com/VirtusLab/git-machete-intellij-plugin/issues/679, isn't it a case anymore?

**Review by mkondratek — APPROVED** [review]

File: frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java (hunk 3 above)
Comment (PawelLipski) [reply frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileViewProviderFactory.java:38]: :nope:, haven't observed so far (I'm testing the version that includes this fix for a few days already)

**Pull request merged**